### PR TITLE
feat(amf): Serialization and Deserialization of Smf Context

### LINKED ISSUE #10924: smf context Serialize and Deserialize

### PR DESCRIPTION
**Pull request by shashidhar-patil**
## Summary
**Serialization and Deserialization of Smf Context to support stateless feature.**

## Test Plan
- Verified basic sanity with ueransim
- Verified using Functional UT
- Please find attached logs and pcap for basic registration

MME logs:
[mme.log](https://github.com/magma/magma/files/7760565/mme.log)

pcap snapshot:
smf_context_pcap

Unit test Snapshot:
smf_context_UT

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.h]: Resolved.

**Comment by shashidhar-patil** [comment]
**PR Rebased.**

mme log: [05/01/2022]
[mme.log](https://github.com/magma/magma/files/7812878/mme.log)

pcap snap: [05/01/2022]
smfcontext_pcap

Unit test: [05/01/2022]
smf_context_2

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: No need to convert to pointer and then dereference.

```suggestion
  snassi_proto->set_sd(state_s_nssai->sd);
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]:
```suggestion
state_s_nssai->sd = snassi_proto.sd();
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Why do we need the pointer creation and dereferencing?
```suggestion
      state_smf_context->gtp_tunnel_id.upf_gtp_teid);
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]:
```suggestion
  state_smf_context->gtp_tunnel_id.upf_gtp_teid =
```

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Here "sd" is a 3-byte raw buffer (uint8_t) but we are storing it as uint32 in protobuff. So this conversion is required. We thought of a couple of approaches and implemented this.
The other approach would be to store sd as "bytes" type in protobuff. But this approach will need conversion to char and then storing it and converting it back while retrieving.
@ssanadhya Please suggest what would be the best approach to this.

Other approach pseudo code(sd is of type "bytes"):
void AmfNasStateConverter::sd_to_chars(const uint8_t* sd, char* sd_array) {
  sd_array[0] = static_cast<char>(sd[0] + ASCII_ZERO);
  sd_array[1] = static_cast<char>(sd[1] + ASCII_ZERO);
  sd_array[2] = static_cast<char>(sd[2] + ASCII_ZERO);
}

void AmfNasStateConverter::chars_to_sd(const char* sd_array, uint8_t* sd) {
  sd[0] = static_cast<int>(sd_array[0]) - ASCII_ZERO;
  sd[1] = static_cast<int>(sd_array[1]) - ASCII_ZERO;
  sd[2] = static_cast<int>(sd_array[2]) - ASCII_ZERO;
}

Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: @shashidhar-patil , you can use the "bytes" type in protobuf and pass the uint8_t array as char array:

State to proto
```
snassi_proto->set_sd((char*) state_s_nssai->sd, SD_LENGTH);
```

Proto to state
```
memcpy((void*)state_s_nssai->sd, (void*)snassi_proto.sd().c_str(), SD_LENGTH);
```

Note that you do not need to convert it into an ASCII value unless you want to print it as text.

Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Thankyou @ssanadhya. Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp]: Addressed this review comment.

**Review by ssanadhya — APPROVED** [review]
LGTM

**Comment by ssanadhya** [comment]
Can be merged once @ardzoht approves.

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**
